src/xshell_base.hpp

```diff
@@ -21,25 +21,25 @@ namespace xeus
 {
     class xserver_zmq_split;
 
-    class xshell
+    class xshell_base
     {
     public:
 
         using listener = std::function<void(xmessage)>;
 
-        xshell(zmq::context_t& context,
+        xshell_base(zmq::context_t& context,
                const std::string& transport,
                const std::string& ip,
                const std::string& shell_port,
                const std::string& sdtin_port,
                xserver_zmq_split* server);
 
-        ~xshell();
+        ~xshell_base();
 
         std::string get_shell_port() const;
         std::string get_stdin_port() const;
 
-        void run();
+        virtual void run();
```

Comment:
As for the hook class, this should be non virtual and forward the call to a private virtual method.